x/stake/types/validator.go

```diff
@@ -383,29 +383,48 @@ func (v Validator) AddTokensFromDel(pool Pool, amount sdk.Int) (Validator, Pool,
 
 	// bondedShare/delegatedShare
 	exRate := v.DelegatorShareExRate()
-	amountDec := sdk.NewDecFromInt(amount)
+	if exRate.IsZero() {
+		panic("zero exRate should not happen")
+	}
 
 	if v.Status == sdk.Bonded {
-		pool = pool.looseTokensToBonded(amountDec)
+		pool = pool.looseTokensToBonded(amount)
 	}
 
-	if exRate.IsZero() {
-		panic("zero exRate should not happen")
-	}
-	v.Tokens = v.Tokens.Add(amountDec)
-	issuedShares := amountDec.Quo(exRate)
+	v.Tokens = v.Tokens.Add(amount)
+	issuedShares := sdk.NewDecFromInt(amount).Quo(exRate)
 	v.DelegatorShares = v.DelegatorShares.Add(issuedShares)
 
 	return v, pool, issuedShares
 }
 
 // RemoveDelShares removes delegator shares from a validator.
-func (v Validator) RemoveDelShares(pool Pool, delShares sdk.Dec) (Validator, Pool, sdk.Dec) {
-	delTokens := v.DelegatorShareExRate().Mul(delShares)
-	delTokens = sdk.MinDec(delTokens, v.Tokens)
-	v, pool = v.RemoveTokens(pool, delTokens)
-	v.DelegatorShares = v.DelegatorShares.Sub(delShares)
-	return v, pool, delTokens
+// NOTE: because token fractions are left in the valiadator,
+//       the exchange rate of future shares of this validator can increase.
+func (v Validator) RemoveDelShares(pool Pool, delShares sdk.Dec) (Validator, Pool, sdk.Int) {
+
+	remainingShares := v.DelegatorShares.Sub(delShares)
+	var issuedTokens sdk.Int
+	if remainingShares.IsZero() {
+
+		// last delegation share gets any trimmings
+		issuedTokens = v.Tokens
+		v.Tokens = sdk.ZeroInt()
+	} else {
+
+		// leave excess tokens in the validator
+		// however fully use all the delegator shares
+		issuedTokens = v.DelegatorShareExRate().Mul(delShares).TruncateInt()
+		issuedTokens = sdk.MinInt(issuedTokens, v.Tokens) // defensive should not occur
```

Comment:
If it shouldn't occur then shouldn't we panic?

Reply:
Yup I agree - will make a panic